### PR TITLE
[inductor] Conditionally copy args to cpu to minimize memory overhead of autotuning

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136701

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by masnesral** [comment]
@eellison (and @shunting314!!):

This is not quite ready for review, but I can quickly clean it up if we feel the approach here is viable. Here's what I've found:

1) Optimizing the copies during auto-tuning does not help for shunting's test_cross_entropy_loss repro like we hoped it would: https://github.com/pytorch/pytorch/pull/129043/files#diff-f8d3282987c90bc3007028a5d7388030ca1c457da0268e8c95a0b366c5354400R147. What I found is that there _is_ one very large tensor copy we can avoid (3GB), but at the point of auto-tuning, the high water mark is already _almost_ 3GB higher than the current memory usage. Therefore, avoiding that copy only saves a few KB.

2) I also tried llm.c using this help from shunting: `python train_gpt2.py --write_tensors=0 --num_iterations=50 --sequence_length=1024 --compile=1 --tensorcores=1 --dtype=bfloat16 --flash=1 --batch_size=32 --input_bin=dev/data/tinyshakespeare/tiny_shakespeare_train.bin --total_batch_size=32768`. Unfortunately, it looks like the number and size of any mutated tensors in that use case is very small so there's just very little opportunity for savings here.

3) I ran this PR through the nightly benchmarks. Results are here: https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Fri%2C%2013%20Sep%202024%2022%3A08%3A45%20GMT&stopTime=Fri%2C%2027%20Sep%202024%2022%3A08%3A45%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/masnesral/118/head&lCommit=6fc14aedee4e93a55457b167672deadcdc350010&rBranch=main&rCommit=db80b98ec460ca5b2fd84c1dfb6426925f64c8cc

From, what I see, very little changes in compile time, performance, memory overhead. EXCEPT there are some fine wins for inference on some timm models: https://hud.pytorch.org/benchmark/timm_models/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Fri,%2013%20Sep%202024%2022:08:45%20GMT&stopTime=Fri,%2027%20Sep%202024%2022:08:45%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/masnesral/118/head&lCommit=6fc14aedee4e93a55457b167672deadcdc350010&rBranch=main&rCommit=db80b98ec460ca5b2fd84c1dfb6426925f64c8cc

You'll notice that the ones with improvements also see a slight degradation in compile time, which makes sense. These results are very close to the theoretical best numbers we saw in the experiment we ran here: https://github.com/pytorch/pytorch/pull/135626

Now some things about the impl:

* Randomizing the offloaded tensors rather than copying from cpu seems to work fine from a perf perspective. But I wonder if it's important to do something more intelligent and try to match the distribution in the original inputs? Randomize taking the min/max into account perhaps?
* Randomizing introduces some non-determinism because the auto-tuning benchmarking loop could introduce rand calls that weren't there before. Furthermore, the _number_ of rand calls could be non-deterministic because the benchmark loop has a portion with time-based cutoff on iterations. Amazingly, I only found one test failure as a result; "fixed" by executing the compile function once to get the auto-tuning out of the way. Not sure how we feel about this downside.
* I skipped randomizing integer types because that somehow seemed more dicey, but I don't know if that even makes any sense.

File: torch/_inductor/runtime/triton_heuristics.py [hunk 1]
```diff
@@ -673,7 +682,67 @@ def kernel_call():
...
+            arg.copy_(cpu_arg, non_blocking=True)
+
+    def prepare_args(self, cpu_copies, *args, **kwargs):
+        from ..compile_fx import clone_preserve_strides
+
+        def prepare_arg(name, arg):
+            if name in cpu_copies:
+                assert torch.is_floating_point(arg)
+                arg.uniform_(0, 1)
+                return arg
```
Comment (eellison) [inline torch/_inductor/runtime/triton_heuristics.py:732]: What was the perf with actually copying back the values instead of randomizing them ? was there a blocker there ? this would be safer. 

See, we just had this ima internally from bad autotuning assumptions https://fb.workplace.com/groups/1075192433118967/posts/1513763445928528. although note, it actually wasnt related to the values of the tensors, but nonetheless, gives me pause on doing something that is not actually safe. 

There's nothing preventing 
```
indices: float

def kernel(tensor, indices):
     tensor[indices.int()]
```

Maybe skipping tensors with indirect indexing is sufficient.

Comment (masnesral) [reply torch/_inductor/runtime/triton_heuristics.py:737]: Compile time was quite bad on my devgpu when copying on every iteration, but let's run the full benchmark suite. I'lll be back...

**Comment by masnesral** [comment]
@eellison I guess the additional compile-time overhead of copying from cpu instead of random doesn't really show up much on these runs (maybe 1s more on timm for the toplevel metric).

Compared to the base commit:
https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Tue%2C%2024%20Sep%202024%2014%3A53%3A06%20GMT&stopTime=Tue%2C%2001%20Oct%202024%2014%3A53%3A06%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/masnesral/118/head&lCommit=d4a837d5caaeac24cf50329568a14f679d5cd7b5&rBranch=main&rCommit=d6d9183456cd07ca0b361a194b98c2fb196e7c36

Compared to the version that uses rand:
https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Tue%2C%2024%20Sep%202024%2014%3A53%3A06%20GMT&stopTime=Tue%2C%2001%20Oct%202024%2014%3A53%3A06%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/masnesral/118/head&lCommit=d4a837d5caaeac24cf50329568a14f679d5cd7b5&rBranch=main&rCommit=d6d9183456cd07ca0b361a194b98c2fb196e7c36

**Review by eellison — COMMENTED** [review]
In our current benchmark code we run eager first, then compile: 

https://github.com/pytorch/pytorch/blob/99eb47fb6de13baf5b4f33dd98e9e02f6166bbbb/benchmarks/dynamo/common.py#L882-L895

Could you do a run where you switch the ordering of these two ? I'm a bit worried that because we hit eager peak memory first we are not replicating how this will be run in the wild.

The full solution would be to do something along the lines of https://github.com/pytorch/pytorch/pull/134874 (or wait til its on by default) and annotate parts of the graph that are at a high water memory or within a clone of mutated args high water memory.

But at the very least, I think we should skip this on the forward of a training loop, where every new node will be saving memory so we clone to cpu on the entire forward.

**Comment by masnesral** [comment]
@eellison 

>  I'm a bit worried that because we hit eager peak memory first we are not replicating how this will be run in the wild

Actually I think we're fine here. Just adding some breakpoints/printfs, it looks like all the autotuning runs inside this warmup function inside the benchmark harness. And while eager runs first, we reset the memory stats in the warmup: https://github.com/pytorch/pytorch/blob/4559cddaf930a8f65f36cffd0ed8a264efad232c/benchmarks/dynamo/common.py#L3169 . I've verified it gets reset before we enter the autotuning.

> The full solution would be to do something along the lines of https://github.com/pytorch/pytorch/pull/134874

Ok sure, tell me more? That looks like a large body of work; a quick look only gives me the high level gist. Do you suggest I dig in to that impl and figure out how to integrate with it?

**Review by eellison — APPROVED** [review]
When I say "full solution" I just mean in terms of annotating the graph for when it is unnecessary to do a copy to cpu. E.g. , early in the graph, even if a clone increases memory, but we know doing the cloen will not increase peak memory.

But from your benchmarking sounds like that's not initially needed.

**Comment by masnesral** [comment]
@eellison mind having another quick look? I added the is_inference / is_backward check

File: torch/_inductor/runtime/triton_heuristics.py [hunk 2]
```diff
@@ -187,6 +187,8 @@ def __init__(
         configs,
         save_cache_hook,
         mutated_arg_names: List[str],  # see [Note: clone mutated buffers]
+        is_inference,
```
Comment (eellison) [inline torch/_inductor/runtime/triton_heuristics.py:190]: In the future we may want to have a more refined check where we will still copy to cuda if we know that it is not a high water mark increasing clone. Can we do that from the start ? I think it will also be more clear inline what this is doing.

File: torch/_inductor/runtime/triton_heuristics.py [hunk 3]
```diff
@@ -702,31 +711,84 @@ def kernel_call():
 
         return benchmarker.benchmark_gpu(kernel_call, rep=40)
 
-    def clone_args(self, *args, **kwargs) -> Tuple[List[Any], Dict[str, Any]]:
-        from ..compile_fx import clone_preserve_strides
+    def copy_args_to_cpu_if_needed(self, *args, **kwargs):
```
Comment (eellison) [inline torch/_inductor/runtime/triton_heuristics.py:714]: We still need to fix `benchmark_fused_nodes` so that we consistently do the same behavior. Can we patch `is_inference` and `is_backward` (or some other way, up to you) so that we always copy arguments inside benchmarking ? see, [here](https://github.com/pytorch/pytorch/blob/0878739b11f95fe092bffc0195d0e7a348612806/torch/_inductor/codegen/triton.py#L3180-L3182)

**Review by eellison — APPROVED** [review]
Looks good - can we make it an explicit field and also fix the benchmarking ? then good to land

File: torch/_inductor/runtime/triton_heuristics.py (hunk 3 above)
Comment (masnesral) [reply torch/_inductor/runtime/triton_heuristics.py:714]: Yeah, we should already be good there if I understand the ask. `benchmark_combo_kernel` and `benchmark_fused_nodes` call `clone_args` directly and I kept the behavior the same there (clone all mutated buffers -- no copying to cpu).

File: torch/_inductor/runtime/triton_heuristics.py [hunk 4]
```diff
@@ -702,31 +709,80 @@ def kernel_call():
...
+        budget = torch.cuda.max_memory_allocated() - torch.cuda.memory_allocated()
+
+        def maybe_copy(name, arg):
+            if name in self.mutated_arg_names:
+                nonlocal budget
                 assert isinstance(arg, torch.Tensor)
-                cloned_args.append(clone_preserve_strides(arg))
-            else:
-                cloned_args.append(arg)
+                assert not arg.is_cpu
```
Comment (int3) [inline torch/_inductor/runtime/triton_heuristics.py:730]: this assert breaks the Triton CPU backend. Should we always disable`optimize_mem` if we are targeting the CPU?

Comment (eellison) [reply torch/_inductor/runtime/triton_heuristics.py:730]: Do we have any tests for the Triton CPU backend? if not, could we forward fix ?

Comment (masnesral) [reply torch/_inductor/runtime/triton_heuristics.py:730]: https://github.com/pytorch/pytorch/pull/137509
